File: integration-tests/src/test/java/stubidp/test/integration/support/StubIdpBuilder.java
package stubidp.test.integration.support;

import stubidp.stubidp.configuration.StubIdp;
import stubidp.stubidp.configuration.UserCredentials;

import java.util.ArrayList;
import java.util.List;

public class StubIdpBuilder {
    private String assetId = "default-stub-idp";
    private String displayName = "Default Stub IDP";
    private String friendlyId = "default-stub-idp";
    private final List<UserCredentials> idpUserCredentials = new ArrayList<>();
    private boolean sendKeyInfo = false;

    public static StubIdpBuilder aStubIdp() {
        return new StubIdpBuilder();
    }

    public StubIdp build() {
        if (idpUserCredentials.isEmpty()) {
            idpUserCredentials.add(new TestUserCredentials("foo", "bar"));
        }

        return new TestStubIdp(assetId, displayName, friendlyId, idpUserCredentials, sendKeyInfo);
    }

    public StubIdpBuilder withId(String id) {
        this.assetId = id;
        this.friendlyId = id;
        return this;
    }

    public StubIdpBuilder withDisplayName(String displayName) {
        this.displayName = displayName;
        return this;
    }

    public StubIdpBuilder addUserCredentials(UserCredentials credentials) {
        this.idpUserCredentials.add(credentials);
        return this;
    }

    public StubIdpBuilder sendKeyInfo(boolean sendKeyInfo) {
        this.sendKeyInfo = sendKeyInfo;
        return this;
    }
}

File: integration-tests/src/test/kotlin/stubidp/kotlin/test/integration/support/StubIdpBuilder.kt
package stubidp.kotlin.test.integration.support

import stubidp.stubidp.configuration.StubIdp
import stubidp.stubidp.configuration.UserCredentials
import stubidp.test.integration.support.TestStubIdp
import stubidp.test.integration.support.TestUserCredentials
import java.util.ArrayList

class StubIdpBuilder {
    private var assetId = "default-stub-idp"
    private var displayName = "Default Stub IDP"
    private var friendlyId = "default-stub-idp"
    private val idpUserCredentials: MutableList<UserCredentials> = ArrayList()
    private var sendKeyInfo = false
    
    fun build(): StubIdp {
        if (idpUserCredentials.isEmpty()) {
            idpUserCredentials.add(TestUserCredentials("foo", "bar"))
        }
        return TestStubIdp(assetId, displayName, friendlyId, idpUserCredentials, sendKeyInfo)
    }

    fun withId(id: String): StubIdpBuilder {
        assetId = id
        friendlyId = id
        return this
    }

    fun withDisplayName(displayName: String): StubIdpBuilder {
        this.displayName = displayName
        return this
    }

    fun addUserCredentials(credentials: UserCredentials): StubIdpBuilder {
        idpUserCredentials.add(credentials)
        return this
    }

    fun sendKeyInfo(sendKeyInfo: Boolean): StubIdpBuilder {
        this.sendKeyInfo = sendKeyInfo
        return this
    }

    companion object {
        @JvmStatic
        fun aStubIdp(): StubIdpBuilder {
            return StubIdpBuilder()
        }
    }
}